kernel/work.c

```diff
@@ -812,6 +812,36 @@ int k_work_queue_unplug(struct k_work_q *queue)
 	return ret;
 }
 
+int k_work_queue_stop(struct k_work_q *queue)
+{
+	__ASSERT_NO_MSG(queue);
+
+	SYS_PORT_TRACING_OBJ_FUNC_ENTER(k_work_queue, stop, queue);
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	if (!flag_test(&queue->flags, K_WORK_QUEUE_STARTED_BIT)) {
+		k_spin_unlock(&lock, key);
+		SYS_PORT_TRACING_OBJ_FUNC_EXIT(k_work_queue, stop, queue, -ENODEV);
+		return -ENODEV;
+	}
+
+	if (!flag_test(&queue->flags, K_WORK_QUEUE_PLUGGED_BIT)) {
+		k_spin_unlock(&lock, key);
+		SYS_PORT_TRACING_OBJ_FUNC_EXIT(k_work_queue, stop, queue, -EBUSY);
+		return -EBUSY;
+	}
+
+	k_thread_abort(&queue->thread);
+	k_thread_join(&queue->thread, K_FOREVER);
```

Comment:
On a technical level: you don't need to do both of these.  k_thread_abort() is already synchronous and will only return when the thread is dead.  Join, at that point, is a guaranteed noop unless something else re-starts the thread from another context, which won't happen because you hold the queue's lock.

But you surely don't want to use abort at all, as it will destroy the target thread in whatever state it might be in, which can be in app code inside callbacks holding resources you can't predict or control.

IMHO the correct technique here is to set an "exit" flag and have the work queue's main loop exit the thread on its own, which you can then reap here with join().

Reply:
And obviously that flag and check could share a lot of implementation with drain()

Reply:
Yepp, agreed.
I'll make the appropriate changes ASAP. 

Thanks for pointing it out.